.buildkite/scripts/steps/cloud/purge_projects.ts

```diff
@@ -0,0 +1,126 @@
+      .map((project) => {
+        const [, prNumber, projectType] = project.name.match(match);
+        return {
+          id: project.id,
+          name: project.name,
+          prNumber,
+          type: projectType,
+        };
+      });
+  } catch (e) {
+    if (e.isAxiosError) {
+      const message = JSON.stringify(e.response.data) || 'unable to fetch projects';
+      throw new Error(message);
+    }
+    throw e;
+  }
+}
+
+async function deleteProject({
+  type,
+  id,
+}: {
+  type: 'elasticsearch' | 'observability' | 'security';
+  id: number;
+}) {
+  try {
+    await projectRequest.delete(`/api/v1/serverless/projects/${type}/${id}`);
+  } catch (e) {
+    if (e.isAxiosError) {
+      const message =
+        JSON.stringify(e.response.data) || `unable to delete ${type} project with id ${id}`;
+      throw new Error(message);
+    }
+    throw e;
+  }
+}
+
+async function purgeProjects() {
+  const prProjects = await getPrProjects();
+  const projectsToPurge = [];
+  for (const project of prProjects) {
+    const NOW = new Date().getTime() / 1000;
+    const DAY_IN_SECONDS = 60 * 60 * 24;
+    const prJson = execSync(
+      `gh pr view '${project.prNumber}' --json state,labels,commits`
+    ).toString();
+    const pullRequest = JSON.parse(prJson);
+    const prOpen = pullRequest.state === 'OPEN';
+    const lastCommit = pullRequest.commits.slice(-1)[0];
+    const lastCommitTimestamp = new Date(lastCommit.committedDate).getTime() / 1000;
+
+    const persistDeployment = Boolean(
+      pullRequest.labels.filter((label: any) => label.name === 'ci:project-persist-deployment')
+        .length
+    );
+    if (prOpen && persistDeployment) {
+      continue;
+    }
+
+    if (!prOpen) {
+      console.log(
+        `Pull Request #${project.prNumber} is no longer open, will delete associated ${project.type} project`
+      );
+      projectsToPurge.push(project);
+    } else if (
+      !Boolean(
+        pullRequest.labels.filter((label: any) =>
+          /^ci:project-deploy-(elasticearch|security|observability)$/.test(label.name)
+        ).length
+      )
+    ) {
+      console.log(
+        `Pull Request #${project.prNumber} no longer has a project deployment label, will delete associated deployment`
+      );
+      projectsToPurge.push(project);
+    } else if (lastCommitTimestamp < NOW - DAY_IN_SECONDS * 2) {
+      console.log(
+        `Pull Request #${project.prNumber} has not been updated in more than 2 days, will delete associated deployment`
+      );
+      projectsToPurge.push(project);
+    }
+  }
+  await Promise.all(projectsToPurge.map((p) => deleteProject(p)));
+}
+
+if (!process.env.PROJECT_API_DOMAIN || !process.env.PROJECT_API_KEY) {
+  console.error('missing project authentication');
+  process.exit(1);
+}
+const projectRequest = axios.create({
+  baseURL: process.env.PROJECT_API_DOMAIN,
+  headers: {
+    Authorization: `ApiKey ${process.env.PROJECT_API_KEY}`,
+  },
+});
+
+try {
+  purgeProjects();
```

Comment:
If this is an async function, this might fail with a promise rejection, and won't be caught by a try-catch

Reply:
Good call, I pushed 3f7a5809c6489a651e10fef4511322121111c743.